### PR TITLE
[Serve] Followup of #42820; Add a DeepSpeed example for HPU 

### PR DESCRIPTION
**Pull request by kira-lin**
## Why are these changes needed?

As discussed offline, we plan to add this DeepSpeed example, without any change in torch_dist.py. In addition, point to llm-on-ray for further examples to add, such as quantization.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 1]
```diff
@@ -158,3 +174,115 @@ in a far-off land, there was a magical kingdom called "Happily Ever Laughter." I
...
+snapshot_download(
+    "meta-llama/Llama-2-70b-chat-hf",
+    # Replace the path if necessary
+    cache_dir=os.getenv("TRANSFORMERS_CACHE", None),
+    # Specify your Hugging Face token
+    token=""
+)
+```
+
+In this example, the deployment replica sends prompts to the Deepspeed workers, which are running in Ray actors:
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:195]:
```suggestion
In this example, the deployment replica sends prompts to the DeepSpeed workers, which are running in Ray actors:
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 2]
```diff
@@ -6,16 +6,28 @@ orphan: true
...
+
+    * How to initialize distributed backend
+
+    * How to load a sharded model onto DeepSpeed workers
+
+    * How to stream responses from DeepSpeed workers
 
 This tutorial helps you serve a large language model on HPUs.
 
+Welcome to try out [llm-on-ray](https://github.com/intel/llm-on-ray), where you can find more examples, and get them deployed in a simpler way.
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:29]: Could you move this to the end of the tutorial in a separate `Next Steps` section? You could say:

```
Check out [llm-on-ray](https://github.com/intel/llm-on-ray) for more ways to customize and deploy LLMs at scale.
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 3]
```diff
@@ -6,16 +6,28 @@ orphan: true
 
 [Habana Gaudi AI Processors (HPUs)](https://habana.ai) are AI hardware accelerators designed by Habana Labs. For more information, see [Gaudi Architecture](https://docs.habana.ai/en/latest/Gaudi_Overview/index.html) and [Gaudi Developer Docs](https://developer.habana.ai/).
 
-This tutorial shows how to deploy [Llama2-7b](https://huggingface.co/meta-llama/Llama-2-7b-chat-hf) using a single HPU:
+This tutorial has two examples:
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:9]:
```suggestion
This tutorial walks through two examples:
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 4]
```diff
@@ -6,16 +6,28 @@ orphan: true
...
 
-* How to load a model in an HPU
+1. Deployment of [Llama2-7b](https://huggingface.co/meta-llama/Llama-2-7b-chat-hf) using a single HPU:
 
-* How to perform generation on an HPU
+    * How to load a model in an HPU
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:13]:
```suggestion
    * How to load a model onto an HPU
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 5]
```diff
@@ -6,16 +6,28 @@ orphan: true
...
 
-* How to enable HPU Graph optimizations
+    * How to perform generation on an HPU
+
+    * How to enable HPU Graph optimizations
+
+2. Deployment of [Llama2-70b](https://huggingface.co/meta-llama/Llama-2-70b-chat-hf) using multiple HPUs on a single node:
+
+    * How to initialize distributed backend
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:21]:
```suggestion
    * How to initialize a distributed backend
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+from ray import serve
+from ray.util.queue import Queue
+from ray.runtime_env import RuntimeEnv
+
+
+@ray.remote(num_cpus=8, resources={"HPU": 1})
+class DeepSpeedInferenceWorker(TorchDistributedWorker):
+    def __init__(self, model_id_or_path: str, world_size: int, local_rank: int):
+        """A worker process of DeepSpeed inference engine.
+        Arguments:
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:19]:
```suggestion
        """An actor that runs a DeepSpeed inference engine.
        
        Arguments:
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+            model_id_or_path: Either an HuggingFace model ID
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:20]:
```suggestion
            model_id_or_path: Either a HuggingFace model ID
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+        # Construct the model with fake meta tensors
+        # It will load the model weights from the checkpoint later
+        with deepspeed.OnDevice(dtype=torch.bfloat16, device="meta"):
+            model = AutoModelForCausalLM.from_config(
+                self.model_config, torch_dtype=torch.bfloat16
+            )
+        model = model.eval()
+
+        # Create a file to indicate where the checkpoint is
+        checkpoints_json = tempfile.NamedTemporaryFile(suffix=".json", mode="+w")
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:78]:
```suggestion
        checkpoints_json = tempfile.NamedTemporaryFile(suffix=".json", mode="w+")
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+        write_checkpoints_json(
+            self.model_id_or_path, self._local_rank, checkpoints_json, token=""
+        )
+
+        # Prepare DeepSpeed inference configuration
+        kwargs = {"dtype": torch.bfloat16}
+        kwargs["checkpoint"] = checkpoints_json.name
+        kwargs["tensor_parallel"] = {"tp_size": self._world_size}
+        # enable hpu graph, similar to cuda graph
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:87]:
```suggestion
        # Enable hpu graph, similar to cuda graph
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+
+    def streaming_generate(self, prompt: str, streamer, **config: Dict[str, Any]):
+        """Generate streamed response given an input."""
+
+        input_ids = self.tokenize(prompt)
+        self.model.generate(input_ids, streamer=streamer, **config)
+
+    def get_streamer(self):
+        """Return a streamer.
+        Only the rank 0 worker's result is needed.
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:116]:
```suggestion
        """Return a streamer.
        
        Only the rank 0 worker's result is needed.
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+        # Get workers' streamers
+        self.streamers = ray.get(
+            [worker.get_streamer.remote() for worker in self.deepspeed_workers]
+        )
+
+    def generate(self, prompt: str, **config: Dict[str, Any]):
+        """Send the prompt to workers for generation.
+        Return after all workers have finished generation.
+        Only return the rank 0 worker's result.
+        """
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:204]:
```suggestion
        """Send the prompt to workers for generation.
        
        Return after all workers have finished generation.
        Only return the rank 0 worker's result.
        """
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+        futures = [
+            worker.generate.remote(prompt, **config)
+            for worker in self.deepspeed_workers
+        ]
+        return ray.get(futures)[0]
+
+    def streaming_generate(self, prompt: str, **config: Dict[str, Any]):
+        """Send the prompt to workers for streaming generation.
+        Only use the rank 0 worker's result.
+        """
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:215]:
```suggestion
        """Send the prompt to workers for streaming generation.
        
        Only use the rank 0 worker's result.
        """
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+        # Enable HPU graph runtime
+        config["hpu_graphs"] = True
+        # Lazy mode should be True when using HPU graphs
+        config["lazy_mode"] = True
+
+        # Non-streaming case
+        if not streaming_response:
+            return self.generate(prompts, **config)
+
+        # streaming case
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:254]:
```suggestion
        # Streaming case
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+import torch
+from transformers import TextStreamer
+
+import ray
+from ray import serve
+from ray.util.queue import Queue
+from ray.runtime_env import RuntimeEnv
+
+
+@ray.remote(num_cpus=8, resources={"HPU": 1})
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:15]: Why 8 CPUs? Could you explain in a comment?

**Review by shrekris-anyscale — COMMENTED** [review]
Nice work overall! I left a few comments– let me know if you have any questions.

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py (hunk 14 above)
Comment (kira-lin) [reply doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py]: It was a random number. I have removed it.

**Review by shrekris-anyscale — APPROVED** [review]
Nice work! @angelinalg Could you take a look?

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+from ray.runtime_env import RuntimeEnv
+
+
+@ray.remote(resources={"HPU": 1})
+class DeepSpeedInferenceWorker:
+    def __init__(self, model_id_or_path: str, world_size: int, local_rank: int):
+        """An actor that runs a DeepSpeed inference engine.
+
+        Arguments:
+            model_id_or_path: Either a HuggingFace model ID
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:21]:
```suggestion
            model_id_or_path: Either a Hugging Face model ID
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+            world_size: Total number of worker processes.
+            local_rank: Rank of this worker process.
+                The rank 0 worker is the head worker.
+        """
+        from transformers import AutoTokenizer, AutoConfig
+        from optimum.habana.transformers.modeling_utils import (
+            adapt_transformers_to_gaudi,
+        )
+
+        # Tweak transformers for better performance on Gaudi
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:32]:
```suggestion
        # Tweak transformers for better performance on Gaudi.
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+        self.device = torch.device("hpu")
+
+        self.model_config = AutoConfig.from_pretrained(
+            model_id_or_path,
+            torch_dtype=torch.bfloat16,
+            token="",
+            trust_remote_code=False,
+        )
+
+        # Load and configure the tokenizer
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:47]:
```suggestion
        # Load and configure the tokenizer.
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+        self.tokenizer = AutoTokenizer.from_pretrained(
+            model_id_or_path, use_fast=False, token=""
+        )
+        self.tokenizer.padding_side = "left"
+        if self.tokenizer.pad_token is None:
+            self.tokenizer.pad_token = self.tokenizer.eos_token
+
+        import habana_frameworks.torch.distributed.hccl as hccl
+
+        # Initialize distributed backend
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:57]:
```suggestion
        # Initialize the distributed backend.
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+        """Load the model to HPU and initialize the DeepSpeed inference engine."""
+
+        import deepspeed
+        from transformers import AutoModelForCausalLM
+        from optimum.habana.checkpoint_utils import (
+            get_ds_injection_policy,
+            write_checkpoints_json,
+        )
+
+        # Construct the model with fake meta tensors
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:73]:
```suggestion
        # Construct the model with fake meta Tensors.
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 20]
```diff
@@ -0,0 +1,272 @@
...
+        # It will load the model weights from the checkpoint later
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:74]:
```suggestion
        # Loads the model weights from the checkpoint later.
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 21]
```diff
@@ -0,0 +1,272 @@
...
+        with deepspeed.OnDevice(dtype=torch.bfloat16, device="meta"):
+            model = AutoModelForCausalLM.from_config(
+                self.model_config, torch_dtype=torch.bfloat16
+            )
+        model = model.eval()
+
+        # Create a file to indicate where the checkpoint is
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:81]:
```suggestion
        # Create a file to indicate where the checkpoint is.
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py (hunk 20 above)
Comment (angelinalg) [reply doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:74]: Not sure if I've changed the meaning of what you're trying to say. If so, please correct accordingly. I wanted to remove the future tense and the ambiguity of "it".

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+        checkpoints_json = tempfile.NamedTemporaryFile(suffix=".json", mode="w+")
+        write_checkpoints_json(
+            self.model_id_or_path, self._local_rank, checkpoints_json, token=""
+        )
+
+        # Prepare DeepSpeed inference configuration
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:87]:
```suggestion
        # Prepare the DeepSpeed inference configuration.
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+        kwargs = {"dtype": torch.bfloat16}
+        kwargs["checkpoint"] = checkpoints_json.name
+        kwargs["tensor_parallel"] = {"tp_size": self._world_size}
+        # Enable hpu graph, similar to cuda graph
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:91]:
```suggestion
        # Enable the HPU graph, similar to the cuda graph.
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 24]
```diff
@@ -0,0 +1,272 @@
...
+        kwargs["enable_cuda_graph"] = True
+        # Specify injection policy, required by DeepSpeed tensor parallelism
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:93]:
```suggestion
        # Specify the injection policy, required by DeepSpeed Tensor parallelism.
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 25]
```diff
@@ -0,0 +1,272 @@
...
+        kwargs["injection_policy"] = get_ds_injection_policy(self.model_config)
+
+        # Initialize the inference engine
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:96]:
```suggestion
        # Initialize the inference engine.
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 26]
```diff
@@ -0,0 +1,272 @@
...
+
+    def generate(self, prompt: str, **config: Dict[str, Any]):
+        """Take in a prompt and generate a response."""
+
+        input_ids = self.tokenize(prompt)
+        gen_tokens = self.model.generate(input_ids, **config)
+        return self.tokenizer.batch_decode(gen_tokens, skip_special_tokens=True)[0]
+
+    def streaming_generate(self, prompt: str, streamer, **config: Dict[str, Any]):
+        """Generate streamed response given an input."""
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:113]:
```suggestion
        """Generate a streamed response given an input."""
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 27]
```diff
@@ -0,0 +1,272 @@
...
+
+        input_ids = self.tokenize(prompt)
+        self.model.generate(input_ids, streamer=streamer, **config)
+
+    def get_streamer(self):
+        """Return a streamer.
+
+        Only the rank 0 worker's result is needed.
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:121]: Can you avoid the passive tense? Who needs only the rank 0 worker's result?

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 28]
```diff
@@ -0,0 +1,272 @@
...
+        if value == self.stop_signal:
+            raise StopIteration()
+        else:
+            return value
+
+
+# __worker_def_end__
+
+# __deploy_def_start__
+# Required variables for DeepSpeed on HPU
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:171]:
```suggestion
# Required variables for DeepSpeed on HPU.
```

Comment (angelinalg) [reply doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:171]: Who requires the variables? If you can remove the passive voice, you could be more precise.

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 29]
```diff
@@ -0,0 +1,272 @@
...
+HABANA_ENVS = {
+    "PT_HPU_LAZY_ACC_PAR_MODE": "0",
+    "PT_HPU_ENABLE_REFINE_DYNAMIC_SHAPES": "0",
+    "PT_HPU_ENABLE_WEIGHT_CPU_PERMUTE": "0",
+    "PT_HPU_ENABLE_LAZY_COLLECTIVES": "true",
+    "HABANA_VISIBLE_MODULES": "0,1,2,3,4,5,6,7",
+}
+
+
+# Define the Ray Serve deployment
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:181]:
```suggestion
# Define the Ray Serve deployment.
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 30]
```diff
@@ -0,0 +1,272 @@
...
+        # Create the DeepSpeed workers
+        self.deepspeed_workers = []
+        for i in range(world_size):
+            self.deepspeed_workers.append(
+                DeepSpeedInferenceWorker.options(
+                    runtime_env=RuntimeEnv(env_vars=HABANA_ENVS)
+                ).remote(model_id_or_path, world_size, i)
+            )
+
+        # Load model in all workers
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:196]:
```suggestion
        # Load the model to all workers.
```

Comment (angelinalg) [reply doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:196]: Not sure if "to" is the idiom. Maybe "onto"?

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 31]
```diff
@@ -0,0 +1,272 @@
...
+        for worker in self.deepspeed_workers:
+            worker.load_model.remote()
+
+        # Get workers' streamers
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:200]:
```suggestion
        # Get the workers' streamers.
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 32]
```diff
@@ -0,0 +1,272 @@
...
+        self.streamers = ray.get(
+            [worker.get_streamer.remote() for worker in self.deepspeed_workers]
+        )
+
+    def generate(self, prompt: str, **config: Dict[str, Any]):
+        """Send the prompt to workers for generation.
+
+        Return after all workers have finished generation.
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:208]:
```suggestion
        Return after all workers finish the generation.
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 33]
```diff
@@ -0,0 +1,272 @@
...
+        streaming_response = json_request["stream"]
+
+        # Prepare prompts
+        prompts = []
+        if isinstance(text, list):
+            prompts.extend(text)
+        else:
+            prompts.append(text)
+
+        # Process config
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:249]:
```suggestion
        # Process the configuration.
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 34]
```diff
@@ -0,0 +1,272 @@
...
+        config.setdefault("max_new_tokens", 128)
+
+        # Enable HPU graph runtime
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:252]:
```suggestion
        # Enable HPU graph runtime.
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 35]
```diff
@@ -0,0 +1,272 @@
...
+        config["hpu_graphs"] = True
+        # Lazy mode should be True when using HPU graphs
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:254]:
```suggestion
        # Lazy mode should be True when using HPU graphs.
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 36]
```diff
@@ -0,0 +1,272 @@
...
+        # Streaming case
+        self.streaming_generate(prompts, **config)
+        return StreamingResponse(
+            self.consume_streamer(self.streamers[0]),
+            status_code=200,
+            media_type="text/plain",
+        )
+
+
+# Replace the model ID with path if necessary
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:270]:
```suggestion
# Replace the model ID with a path if necessary.
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 37]
```diff
@@ -6,16 +6,28 @@ orphan: true
 
 [Habana Gaudi AI Processors (HPUs)](https://habana.ai) are AI hardware accelerators designed by Habana Labs. For more information, see [Gaudi Architecture](https://docs.habana.ai/en/latest/Gaudi_Overview/index.html) and [Gaudi Developer Docs](https://developer.habana.ai/).
```
Comment (angelinalg) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:7]:
```suggestion
[Habana Gaudi AI Processors (HPUs)](https://habana.ai) are AI hardware accelerators designed by Habana Labs. See [Gaudi Architecture](https://docs.habana.ai/en/latest/Gaudi_Overview/index.html) and [Gaudi Developer Docs](https://developer.habana.ai/) for more details.
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 38]
```diff
@@ -6,16 +6,28 @@ orphan: true
...
 
-This tutorial shows how to deploy [Llama2-7b](https://huggingface.co/meta-llama/Llama-2-7b-chat-hf) using a single HPU:
+This tutorial walks through two examples:
```
Comment (angelinalg) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:9]:
```suggestion
This tutorial has two examples:
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 39]
```diff
@@ -6,16 +6,28 @@ orphan: true
...
 
-* How to load a model in an HPU
+1. Deployment of [Llama2-7b](https://huggingface.co/meta-llama/Llama-2-7b-chat-hf) using a single HPU:
 
-* How to perform generation on an HPU
+    * How to load a model onto an HPU
```
Comment (angelinalg) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:13]:
```suggestion
    * Load a model onto an HPU.
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 40]
```diff
@@ -6,16 +6,28 @@ orphan: true
...
 
-* How to enable HPU Graph optimizations
+    * How to perform generation on an HPU
```
Comment (angelinalg) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:15]:
```suggestion
    * Perform generation on an HPU.
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 41]
```diff
@@ -6,16 +6,28 @@ orphan: true
...
+
+    * How to enable HPU Graph optimizations
```
Comment (angelinalg) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:17]:
```suggestion
    * Enable HPU Graph optimizations.
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 42]
```diff
@@ -6,16 +6,28 @@ orphan: true
...
+
+2. Deployment of [Llama2-70b](https://huggingface.co/meta-llama/Llama-2-70b-chat-hf) using multiple HPUs on a single node:
+
+    * How to initialize a distributed backend
```
Comment (angelinalg) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:21]:
```suggestion
    * Initialize a distributed backend.
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 43]
```diff
@@ -6,16 +6,28 @@ orphan: true
...
+
+    * How to load a sharded model onto DeepSpeed workers
```
Comment (angelinalg) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:23]:
```suggestion
    * Load a sharded model onto DeepSpeed workers.
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 44]
```diff
@@ -6,16 +6,28 @@ orphan: true
...
+
+    * How to stream responses from DeepSpeed workers
```
Comment (angelinalg) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:25]:
```suggestion
    * Stream responses from DeepSpeed workers.
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 45]
```diff
@@ -6,16 +6,28 @@ orphan: true
...
 
 This tutorial helps you serve a large language model on HPUs.
```
Comment (angelinalg) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:27]:
```suggestion
This tutorial serves a large language model (LLM) on HPUs.
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 46]
```diff
@@ -6,16 +6,28 @@ orphan: true
...
 
+
+
 ## Environment setup
 
 We recommend using a prebuilt container to run these examples. To run a container, you need Docker. See [Install Docker Engine](https://docs.docker.com/engine/install/) for installation instructions.
```
Comment (angelinalg) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:33]:
```suggestion
Use a prebuilt container to run these examples. To run a container, you need Docker. See [Install Docker Engine](https://docs.docker.com/engine/install/) for installation instructions.
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 47]
```diff
@@ -78,6 +90,10 @@ To follow the examples in this tutorial, mount the directory containing the exam
 ```bash
 pip install ray[tune,serve]
 pip install git+https://github.com/huggingface/optimum-habana.git
+# Replace 1.14.0 with the driver version of the container
```
Comment (angelinalg) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:93]:
```suggestion
# Replace 1.14.0 with the driver version of the container.
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 48]
```diff
@@ -78,6 +90,10 @@ To follow the examples in this tutorial, mount the directory containing the exam
...
+pip install git+https://github.com/HabanaAI/DeepSpeed.git@1.14.0
+# Only needed by the DeepSpeed example
```
Comment (angelinalg) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:95]:
```suggestion
# Only needed by the DeepSpeed example.
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 49]
```diff
@@ -158,3 +174,118 @@ in a far-off land, there was a magical kingdom called "Happily Ever Laughter." I
 In this kingdom, there was a beautiful princess named Lily. She was kind, gentle, and had a heart full of laughter. Every day, she would wake up with a smile on her face, ready to face whatever adventures the day might bring.
 One day, a wicked sorcerer cast a spell on the kingdom, causing all
 ```
+
+## Running a sharded model on multiple HPUs
+
+This example shows how to deploy a Llama2-70b model using 8 HPUs orchestrated by DeepSpeed.
```
Comment (angelinalg) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:180]:
```suggestion
This example deploys a Llama2-70b model using 8 HPUs orchestrated by DeepSpeed. 
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 50]
```diff
@@ -158,3 +174,118 @@ in a far-off land, there was a magical kingdom called "Happily Ever Laughter." I
...
+
+This example shows how to deploy a Llama2-70b model using 8 HPUs orchestrated by DeepSpeed. 
+
+The example requires caching the Llama2-70b model. Run the following Python code in the Habana container to cache the model. 
+
+```python
+from huggingface_hub import snapshot_download
+snapshot_download(
+    "meta-llama/Llama-2-70b-chat-hf",
+    # Replace the path if necessary
```
Comment (angelinalg) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:188]:
```suggestion
    # Replace the path if necessary.
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 51]
```diff
@@ -158,3 +174,118 @@ in a far-off land, there was a magical kingdom called "Happily Ever Laughter." I
...
+    cache_dir=os.getenv("TRANSFORMERS_CACHE", None),
+    # Specify your Hugging Face token
```
Comment (angelinalg) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:190]:
```suggestion
    # Specify your Hugging Face token.
```

File: doc/source/serve/tutorials/intel-habana-gaudi-inference.md [hunk 52]
```diff
@@ -158,3 +174,118 @@ in a far-off land, there was a magical kingdom called "Happily Ever Laughter." I
...
+The path was overgrown with weeds and vines, and it looked as though it hadn't been traversed in years. But Sophia was intrigued, and she decided to follow it to see where it led. She pushed aside the branches and stepped onto the path
+Once upon a time, there was a young woman named Sophia who lived in a small village nestled in the rolling hills of Tuscany. Sophia was a curious and adventurous soul, always eager to explore the world around her. One day, while wandering through the village, she stumbled upon a hidden path she had never seen before.
+The path was overgrown with weeds and vines, and it looked as though it hadn't been traversed in years. But Sophia was intrigued, and she decided to follow it to see where it led. She pushed aside the branches and stepped onto the path
+```
+
+## Next Steps
+Check out [llm-on-ray](https://github.com/intel/llm-on-ray) for more ways to customize and deploy LLMs at scale.
```
Comment (angelinalg) [inline doc/source/serve/tutorials/intel-habana-gaudi-inference.md:291]:
```suggestion
See [llm-on-ray](https://github.com/intel/llm-on-ray) for more ways to customize and deploy LLMs at scale.
```

**Review by angelinalg — APPROVED** [review]
Nice job! I have some nits for style consistency, if you wouldn't mind. Please feel free to reach out if you have questions.

**Pull request merged**